### PR TITLE
feat(dbus): reset firewalld to default settings

### PR DESCRIPTION
**Pull request by BrennanPaciorek**
OVERVIEW

- Allows users to reset firewalld to its clean slate
  - option in client and clis (not in applet)
- implemented via a new method firewalld.resetToDefaults()
  - in org.fedoraproject.FirewallD1

FILE CHANGES
src/firewall/server/firewalld.py
- dbus method reset() - removes custom config from firewalld and reloads
src/firewall/core/fw_config.py
 - method reset_defaults()
   - reset_defaults()
     - resets permanent configuration to default
     - used by both online and offline cli

src/firewall/client.py
 - FirewallClient.resetToDefaults() - interface to firewalld.resetToDefaults()

src/firewall-cmd.in
 - `--reset-to-defaults` standalone option
   - calls client's resetToDefaults()

src/firewall-offline-cmd
 - `--reset-to-defaults` standalone option
   - calls fw_config.reset_defaults() to reset permanent configuration

src/tests/features/reset_defaults.at
 - new file - tests for reset-to-defaults cli option

src/tests/features/features.at
 - now includes reset_defaults.at tests

src/tests/functions.at
 - modify FWD_CHECK to allow firewall-offline-cmd tests to run properly
   in case of `--reset-to-defaults` testing

shell-completion/bash/firewall-cmd
 - `--reset-to-defaults` argument added

docs/xml/firewalld.dbus.xml
 - documentation for resetToDefaults()

docs/xml/firewall-cmd.xml.in
 - documentation for `--reset-to-defaults`

docs/xml/firewall-offline-cmd.xml
 - documentation for offline `--reset-to-defaults` option

### CODE REVIEW

**Comment by BrennanPaciorek** [comment]
that PR summary could use some work, for a more readable one, refer to the first (only) commit message for the branch

File: src/firewall-cmd.in
Comment (BrennanPaciorek) [inline src/firewall-cmd.in:65]: Leave `--reset-to-defaults` in general? additionally, should it be `--reset-to-defaults` or `--reset-defaults`?

File: src/firewall-cmd.in
Comment (BrennanPaciorek) [inline src/firewall-cmd.in:2490]: Decided to reload here instead of at fw_config.py so I do not need to check if running in offline mode, could add a parameter not to attempt to reload though to keep the main logic contained to firewall/core.

File: src/firewall/core/fw_config.py
Comment (BrennanPaciorek) [inline src/firewall/core/fw_config.py:186]: as mentioned, should I introduce a parameter to check if the method is running in offline mode? The only valid places would be here or the cli to get access to FirewallD's reload() functionality, unless resetDefaults() (client.py) was moved to FirewallClient

File: src/firewall/server/config.py
Comment (BrennanPaciorek) [inline src/firewall/server/config.py:1780]: can probably just move this in with old_config dict

File: src/firewall/server/config.py
Comment (BrennanPaciorek) [inline src/firewall/server/config.py:1801]: move all this to another helper method?

File: src/firewall/server/config.py
Comment (BrennanPaciorek) [inline src/firewall/server/config.py:1785]: reload unnecessary here if I am reloading from cli

**Review by BrennanPaciorek — COMMENTED** [review]
TODO: more coverage of feature testing, find way to test restoring old settings in the event of an exception.

File: src/firewall/core/fw_config.py
Comment (erig0) [inline src/firewall/core/fw_config.py:249]: It's a lot of work to sort out these io_objects and find the correct remove function. Since we're going to reload can we simply remove/unlink the XML files in `/etc/firewalld/{zones,policies,services,ipsets,icmptypes,helpers,etc}`?

File: src/firewall/client.py
Comment (erig0) [inline src/firewall/client.py:3128]: Please call it `reset_to_defaults`, i.e. keep the naming consistent.

Comment (BrennanPaciorek) [reply src/firewall/client.py]: Fixed, also updated reset() in server/firewalld.py.

File: src/firewall/core/fw_config.py
Comment (BrennanPaciorek) [reply src/firewall/core/fw_config.py]: I have now updated it to just remove the xml files in the same way as the type specific methods would. Working to find a good workaround for {set,new}_io_object() methods for restoring settings in the event of reset's failure.

**Comment by erig0** [comment]
> TODO: test restoring old settings in the event of an exception.

I don't think handling that case is worth the code complexity. If unlinking files fails, then it's unlikely re-writing them is going to succeed. Just let it fail.

**Comment by BrennanPaciorek** [comment]
> I don't think handling that case is worth the code complexity. If unlinking files fails, then it's unlikely re-writing them is going to succeed. Just let it fail.

Removed code pertaining to restoring settings on fail.

**Review by erig0 — APPROVED** [review]
Well done Brennan. I changed some wording (s/version//). But it looks good now. Just waiting on CI before I merge.

**Pull request merged**
